### PR TITLE
Hessian

### PR DESCRIPTION
**Pull request by softmattertheory**
Improvements to linear algebra on morpho

### CODE REVIEW

File: morpho5/datastructures/matrix.c [hunk 1]
```diff
@@ -949,6 +1042,106 @@ value Matrix_norm(vm *v, int nargs, value *args) {
     return MORPHO_FLOAT(matrix_norm(a));
 }
 
+/** Matrix eigensystem */
+bool Matrix_eigen(vm *v, objectmatrix *a, value *evals, value *evecs) {
```
Comment (joshichaitanya3) [inline morpho5/datastructures/matrix.c:1046]: This is not a Morpho method. Should the M be lowercase?

File: morpho5/datastructures/matrix.c [hunk 2]
```diff
@@ -1052,6 +1265,9 @@ MORPHO_METHOD(MATRIX_INNER_METHOD, Matrix_inner, BUILTIN_FLAGSEMPTY),
 MORPHO_METHOD(MORPHO_SUM_METHOD, Matrix_sum, BUILTIN_FLAGSEMPTY),
 MORPHO_METHOD(MATRIX_NORM_METHOD, Matrix_norm, BUILTIN_FLAGSEMPTY),
 MORPHO_METHOD(MATRIX_TRANSPOSE_METHOD, Matrix_transpose, BUILTIN_FLAGSEMPTY),
+MORPHO_METHOD(MATRIX_RESHAPE_METHOD, Matrix_reshape, BUILTIN_FLAGSEMPTY),
+MORPHO_METHOD(MATRIX_EIGENVALUES_METHOD, Matrix_eigenvalues, BUILTIN_FLAGSEMPTY),
+MORPHO_METHOD(MATRIX_EIGENSYSTEM_METHOD, Matrix_eigensystem, BUILTIN_FLAGSEMPTY),
```
Comment (joshichaitanya3) [inline morpho5/datastructures/matrix.c:1270]: `reshape`, `eigenvalues` and `eigensystem` methods need documentation

File: morpho5/datastructures/matrix.c [hunk 3]
```diff
@@ -532,11 +598,21 @@ value matrix_constructor(vm *v, int nargs, value *args) {
     } else if (nargs==1 &&
                MORPHO_ISLIST(MORPHO_GETARG(args, 0))) {
         new=object_matrixfromlist(MORPHO_GETLIST(MORPHO_GETARG(args, 0)));
-        if (!new) morpho_runtimeerror(v, MATRIX_INVLDARRAYINIT);
+        if (!new) {
+            /** Could this be a concatenation operation? */
```
Comment (joshichaitanya3) [inline morpho5/datastructures/matrix.c:602]: Matrix concatenation needs documentation

File: morpho5/geometry/field.c [hunk 4]
```diff
@@ -823,7 +830,8 @@ MORPHO_METHOD(FIELD_OP_METHOD, Field_op, BUILTIN_FLAGSEMPTY),
 MORPHO_METHOD(MORPHO_PRINT_METHOD, Field_print, BUILTIN_FLAGSEMPTY),
 MORPHO_METHOD(MORPHO_CLONE_METHOD, Field_clone, BUILTIN_FLAGSEMPTY),
 MORPHO_METHOD(FIELD_SHAPE_METHOD, Field_shape, BUILTIN_FLAGSEMPTY),
-MORPHO_METHOD(FIELD_MESH_METHOD, Field_mesh, BUILTIN_FLAGSEMPTY)
+MORPHO_METHOD(FIELD_MESH_METHOD, Field_mesh, BUILTIN_FLAGSEMPTY),
+MORPHO_METHOD(FIELD_LINEARIZE_METHOD, Field_linearize, BUILTIN_FLAGSEMPTY)
```
Comment (joshichaitanya3) [inline morpho5/geometry/field.c:834]: `linearize` and `mesh` aren't documented

**Review by joshichaitanya3 — CHANGES_REQUESTED** [review]
Looks good, I am quite excited to use this!
The new methods for Matrix and Field need documentation. Added some other minor comments. 

There is some merge conflict that needs to be resolved.

File: morpho5/datastructures/matrix.c (hunk 1 above)
Comment (softmattertheory) [reply morpho5/datastructures/matrix.c]: Good spot! I fixed the capitalization.

File: morpho5/datastructures/matrix.c (hunk 2 above)
Comment (softmattertheory) [reply morpho5/datastructures/matrix.c:1291]: Done, and I actually added a lot of additional documentation for the Matrix class too.

File: morpho5/geometry/field.c (hunk 4 above)
Comment (softmattertheory) [reply morpho5/geometry/field.c:836]: I documented mesh. Linearize is still a bit experimental, so won't document yet.

File: morpho5/geometry/functional.c [hunk 5]
```diff
@@ -3086,7 +3281,15 @@ bool areaintegral_integrand(vm *v, objectmesh *mesh, elementid id, int nv, int *
     if (fabs(nnorm)>MORPHO_EPS) functional_vecscale(mesh->dim, 1.0/nnorm, normaldata, normaldata);
     objectmatrix mnormal = MORPHO_STATICMATRIX(normaldata, mesh->dim, 1);
     norml = MORPHO_OBJECT(&mnormal);
-    
+
+    /* Evaluate gradient */
+    /* -- Temporary code below -- */
```
[the remark this answers is not in the document]
Comment (softmattertheory) [reply morpho5/geometry/functional.c:3315]: For now; it's temporary code that I was using for a particular problem. There needs to be a more general implementation in the next version.

**Pull request merged**